arch/xtensa/src/esp32/esp32_spiflash.c

```diff
@@ -862,108 +974,380 @@ static int esp32_writedata(FAR struct esp32_spiflash_s *priv, uint32_t addr,
  *
  * Input Parameters:
  *   spi    - ESP32 SPI Flash chip data
  *   addr   - target address
  *   buffer - data buffer pointer
- *   size   - data number
+ *   size   - data number by bytes
  *
  * Returned Value:
  *   OK if success or a negative value if fail.
  *
  ****************************************************************************/
 
-static int esp32_readdata(FAR struct esp32_spiflash_s *priv, uint32_t addr,
-                          uint8_t *buffer, uint32_t size)
+static int IRAM_ATTR esp32_readonce(FAR struct esp32_spiflash_s *priv,
+                                    uint32_t addr,
+                                    uint32_t *buffer,
+                                    uint32_t size)
 {
   uint32_t regval;
   uint32_t i;
-  uint32_t bytes;
-  uint32_t tmp;
-  uint32_t res;
-  uint8_t  *tmpbuff = buffer;
-  int me;
-  uint32_t flags;
+
+  if (size > SPI_FLASH_READ_BUF_SIZE)
+    {
+      return -EINVAL;
+    }
 
   if (esp32_wait_idle(priv) != OK)
     {
       return -EIO;
     }
 
-  while (size > 0)
+  regval = ((size << 3) - 1) << SPI_USR_MISO_DBITLEN_S;
+  spi_set_reg(priv, SPI_MISO_DLEN_OFFSET, regval);
+
+  regval = addr << 8;
+  spi_set_reg(priv, SPI_ADDR_OFFSET, regval);
+
+  spi_set_reg(priv, SPI_RD_STATUS_OFFSET, 0);
+  spi_set_reg(priv, SPI_CMD_OFFSET, SPI_USR);
+  while (spi_get_reg(priv, SPI_CMD_OFFSET) != 0)
     {
-      flags = esp32_spiflash_opstart(priv, &me);
+      ;
+    }
+
+  for (i = 0; i < (size / 4); i++)
+    {
+      buffer[i] = spi_get_reg(priv, SPI_W0_OFFSET + i * 4);
+    }
+
+  if (size & 0x3)
+    {
+      regval = spi_get_reg(priv, SPI_W0_OFFSET + i * 4);
+      memcpy(&buffer[i], &regval, size & 0x3);
+    }
+
+  return OK;
+}
 
+/****************************************************************************
+ * Name: esp32_readdata
+ *
+ * Description:
+ *   Read data from SPI Flash at designated address.
+ *
+ * Input Parameters:
+ *   spi    - ESP32 SPI Flash chip data
+ *   addr   - target address
+ *   buffer - data buffer pointer
+ *   size   - data number
+ *
+ * Returned Value:
+ *   OK if success or a negative value if fail.
+ *
+ ****************************************************************************/
+
+static int IRAM_ATTR esp32_readdata(FAR struct esp32_spiflash_s *priv,
+                                    uint32_t addr,
+                                    uint8_t *buffer,
+                                    uint32_t size)
+{
+  int ret;
+  int me;
+  uint32_t flags;
+  uint32_t off = 0;
+  uint32_t bytes;
+  uint32_t tmp_buf[SPI_FLASH_READ_WORDS];
+
+  while (size > 0)
+    {
       bytes = MIN(size, SPI_FLASH_READ_BUF_SIZE);
-      regval = ((bytes << 3) - 1) << SPI_USR_MISO_DBITLEN_S;
-      spi_set_reg(priv, SPI_MISO_DLEN_OFFSET, regval);
 
-      regval = addr << 8;
-      spi_set_reg(priv, SPI_ADDR_OFFSET, regval);
+      flags = esp32_spiflash_opstart(priv, &me);
+      ret = esp32_readonce(priv, addr, tmp_buf, bytes);
```

Comment:
`esp32_spiflash_opstart` will disable the cache and `esp32_readonce` is using standard `memcpy`.

Reply:
I have added option to select `LIB_ARCH_MEMCPY`  when selecting ARCH_CHIP_ESP32. https://github.com/apache/incubator-nuttx/blob/b63c0863b27d1dd884a47594fa6a59f203de47eb/arch/xtensa/Kconfig#L21

Reply:
Is it in IRAM?

Reply:
When selecting `LIBC_ARCH_MEMCPY `, Nuttx will disable standard `memcpy` and use ROM's `memcpy`, then it will work as esp-idf.